Sources/SwiftKafka/KafkaPollingSystem.swift

```diff
@@ -0,0 +1,459 @@
+    /// - Returns: An awaitable task representing the execution of the poll loop.
+    func run(pollInterval: Duration) async throws {
+        switch self.stateMachine.withLockedValue({ $0.run() }) {
+        case .alreadyClosed:
+            throw KafkaError.pollLoop(reason: "Invocation of \(#function) failed, poll loop already closed.")
+        case .alreadyRunning:
+            fatalError("Poll loop must not be started more than once")
+        case .startLoop:
+            break
+        }
+
+        while true {
+            let action = self.nextPollLoopAction()
+
+            switch action {
+            case .pollAndSleep(let pollClosure):
+                // Poll Kafka for new acknowledgements and sleep for the given
+                // pollInterval to avoid hot looping.
+                pollClosure?()
+                do {
+                    try await Task.sleep(for: pollInterval)
+                } catch {
+                    self.terminate()
+                    throw error
+                }
+            case .suspendPollLoop:
+                // The downstream consumer asked us to stop sending new messages.
+                // We therefore await until we are unsuspended again.
+                try await withTaskCancellationHandler {
+                    // Necessary.
+                    // We don't want to create a leaking continuation
+                    // when we are already cancelled.
+                    // The cancellation of a Task does not mean
+                    // its code will not still be executed.
+                    guard Task.isCancelled == false else {
+                        self.terminate(CancellationError())
+                        throw CancellationError()
+                    }
+                    try await withCheckedThrowingContinuation { continuation in
+                        self.stateMachine.withLockedValue { $0.suspendLoop(continuation: continuation) }
+                    }
+                } onCancel: {
+                    self.terminate(CancellationError())
+                    return
+                }
+            case .shutdownPollLoop:
+                // We have been asked to close down the poll loop.
+                self.terminate()
+                return
+            }
+        }
+    }
+
+    /// Yield new elements to the underlying `NIOAsyncSequenceProducer`.
+    func yield(_ element: Element) {
+        let action = self.stateMachine.withLockedValue { $0.yield(element) }
+        switch action {
+        case .produceMore:
+            self.produceMore()
+        case .stopProducing:
+            self.stopProducing()
+        case .terminate:
+            self.terminate()
+        case .none:
+            break
+        }
+    }
+
+    /// Determines the next action to be taken in the poll loop based on the current state.
+    ///
+    /// - Returns: The next action for the poll loop.
+    func nextPollLoopAction() -> KafkaPollingSystem.StateMachine.PollLoopAction {
```

Comment:
Do we really need a separate method for this?

Reply:
Not really, but given that our `stateMachine` is `private` this provides a safe way to access the poll loop's state for testing. There are not too many things we can assert on here in testing. Alternatively we can make `KafkaPollingSystem.yield(_:)` return the next action but this would also then just serve the purpose on being able to assert on the `KafkaPollingSystem`. In general all of this is quite hard as the `KafkaPollingSystem` does not expose that much information to the outside world.